core/arch/arm/kernel/secure_partition.c

```diff
@@ -1,50 +1,405 @@
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sp_session *s = NULL;
+	struct sp_ctx *ctx = NULL;
+
+	if (!find_secure_partition(uuid))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	res = sp_create_session(open_sessions, uuid, &s);
+	if (res != TEE_SUCCESS) {
+		DMSG("init sp session failed %#"PRIx32, res);
+		return res;
+	}
+
+	if (s->ts_sess.ctx)
+		ctx = to_sp_ctx(s->ts_sess.ctx);
+
+	if (!ctx)
+		return TEE_ERROR_TARGET_DEAD;
+	*sess = s;
+
+	ts_push_current_session(&s->ts_sess);
+	/* Load the SP using ldelf. */
+	ldelf_load_ldelf(&ctx->uctx);
+	res = ldelf_init_with_ldelf(&s->ts_sess, &ctx->uctx);
+
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed. loading SP using ldelf %#"PRIx32, res);
+		return TEE_ERROR_TARGET_DEAD;
+	}
+
+	/*Make the SP ready for its first run*/
+	s->state = sp_idle;
+	s->caller_id = 0;
+	sp_init_set_registers(ctx);
+	ts_pop_current_session();
+
+	if (res != TEE_SUCCESS)
+		EMSG("SP loading failed with error %#"PRIx32, res);
+
+	return res;
+}
+
+static TEE_Result sp_init_uuid(const TEE_UUID *uuid)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sp_session *sess = NULL;
+	struct thread_smc_args args = {0};
+
+	res = sp_open_session(&sess,
+			      &open_sp_sessions,
+			      uuid);
+
+	ts_push_current_session(&sess->ts_sess);
+	sp_init_info(to_sp_ctx(sess->ts_sess.ctx), &args);
+	ts_pop_current_session();
+
+	if (sp_enter(&args, sess))
+		return FFA_ABORTED;
+
+	spmc_sp_msg_handler(&args, sess);
+
+	return res;
+}
+
+TEE_Result sp_enter(struct thread_smc_args *args, struct sp_session *sp)
+{
+	TEE_Result res = FFA_OK;
+	struct sp_ctx *ctx = to_sp_ctx(sp->ts_sess.ctx);
+
+	ctx->sp_regs.x[0] = args->a0;
+	ctx->sp_regs.x[1] = args->a1;
+	ctx->sp_regs.x[2] = args->a2;
+	ctx->sp_regs.x[3] = args->a3;
+	ctx->sp_regs.x[4] = args->a4;
+	ctx->sp_regs.x[5] = args->a5;
+	ctx->sp_regs.x[6] = args->a6;
+	ctx->sp_regs.x[7] = args->a7;
+
+	res = sp->ts_sess.ctx->ops->enter_invoke_cmd(&sp->ts_sess, 0);
+
+	args->a0 = ctx->sp_regs.x[0];
+	args->a1 = ctx->sp_regs.x[1];
+	args->a2 = ctx->sp_regs.x[2];
+	args->a3 = ctx->sp_regs.x[3];
+	args->a4 = ctx->sp_regs.x[4];
+	args->a5 = ctx->sp_regs.x[5];
+	args->a6 = ctx->sp_regs.x[6];
+	args->a7 = ctx->sp_regs.x[7];
+
+	return res;
+}
+
+static TEE_Result sp_enter_invoke_cmd(struct ts_session *s,
+				      uint32_t cmd __unused)
+{
+	struct sp_ctx *ctx = to_sp_ctx(s->ctx);
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t exceptions = 0;
+	uint64_t cpsr = 0;
+	struct sp_session *sp_s = to_sp_session(s);
+	struct ts_session *sess = NULL;
+	struct thread_ctx_regs *sp_regs = NULL;
+	uint32_t panicked = false;
+	uint32_t panic_code = 0;
+
+	bm_timestamp();
+
+	sp_regs = &ctx->sp_regs;
+	ts_push_current_session(s);
+
+	cpsr = sp_regs->cpsr;
+	sp_regs->cpsr = read_daif() & (SPSR_64_DAIF_MASK << SPSR_64_DAIF_SHIFT);
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+
+	__thread_enter_user_mode(sp_regs, &panicked, &panic_code);
+	sp_regs->cpsr = cpsr;
+	thread_unmask_exceptions(exceptions);
+
+#ifdef CFG_WITH_VFP
```

Comment:
can remove this line and always call `thread_user_clear_vfp();`.

Reply:
done